### PR TITLE
Fixed #33879 -- Fixed wrong results for long intervals.

### PR DESCRIPTION
**Pull request by GianpaoloBranca**
Added 10 hours to the month duration for intervals longer than 2 months.
Adding 10 hours directly to the constant would break other tests,
so the fix is added in the algorithm and only affects long intervals.

ticket-33879

### CODE REVIEW

File: django/utils/timesince.py [hunk 1]
```diff
@@ -88,6 +88,9 @@ def timesince(d, now=None, reversed=False, time_strings=None, depth=2):
         if count == 0:
             break
         result.append(avoid_wrapping(time_strings[name] % {"num": count}))
+        # prevent errors in long intervals
+        if name == "month" and count > 2:
+            seconds += 10 * 60 * 60
```
Comment (carltongibson) [inline django/utils/timesince.py:93]: I think we need more tests to show this is correct for all values of count…  It's not clear to me why a single increment of 10 hours would be equally value if count is 2  or if it's 11.  

One of the initial ideas on the ticket was to make the `seconds` divisor relative to a 30.4 days, based on a a first pass at `count`. That may be too complex. But it would sum so the added `.4` would be greater as `count` increased. Don't we need something like that? 🤔

I'd suggest adding tests for just before and after each month increment to make sure we got the right result. e.g. Jan 1 to Mar 1 is 2 months, ..., Jan 1 to Dec 29 is 11 months 4 weeks, etc.. (These could be in a single case using `subTest()`.)

**Review by carltongibson — COMMENTED** [review]
Hi @GianpaoloBranca — did you look at all at @uri-rodberg's issue from [comment:4 on the ticket](https://code.djangoproject.com/ticket/33879#comment:4) ref `1\u00A0year, 12\u00A0months ago` output? Would you plan to address that too? 

I think I'd like to see a few more test cases around the edge-cases here: to show that we get the right result at the change points. What do you think? 🤔

File: django/utils/timesince.py (hunk 1 above)
Comment (GianpaoloBranca) [reply django/utils/timesince.py:93]: Hi @carltongibson, 

Thanks for reviewing my patch. This is my first contribution

The reason for 10 hours was that 10 hours are 0.4167 days, and the average month duration is 365/12 = 30.4167...so the average month is exactly 30 days and 10 hours long.

The main issue here is that there is a `month` constant but months do not have a constant value... I just did a quick test and for the delta from Jan 1 to Mar 1 the result is not 2 months, but 1 month and 4 weeks (which btw, could be considered at least technically correct).

I am now working on a fix that produces the most humanly correct result possible, but to do so I would need to modify some of the pre-exisisting tests as well. There are some tests in `humanize_tests.tests.HumanizeTests` and `utils_tests.test_timesince.TimesinceTests` that needs to be updated. 

For example, there is one test where:
- `d` is 2012/3/9 
- `now` is `d` plus 30 days, that is 2012/4/8

which is not yet one month but the test expects the result to be one month. The correct result would be 4 weeks and 2 days. I was thinking of fixing this by changing the delta from 30 to 31 days for this test, and implement similar fixes in other tests that may fail for the same reason.

Should I modify these tests?

**Comment by GianpaoloBranca** [comment]
I believe the fix is now ready and tested for all the edge cases.

The algorithm calculates the average month duration from `d` to `now` , and set that value in `timesince_chunks_adjusted` (I can't update directly the `TIMESINCE_CHUNKS`), that it is used in the chunks loop. 

The part in the loop with:
```
if name == "month":
    count = min(count, months)
```
Is necessary, as there are cases where `count` ends up being greater than `months`, and this would be incorrect.

@carltongibson can you please have a look?

File: django/utils/timesince.py [hunk 2]
```diff
@@ -69,22 +71,38 @@ def timesince(d, now=None, reversed=False, time_strings=None, depth=2):
             leapdays += 1
     delta -= datetime.timedelta(leapdays)
 
+    # fix months duration to be relative to the months between d and now
+    months = now.month - d.month
+    if months > 0:
+        m_days = (sum(MONTHS_DAYS[d.month - 1 : now.month - 1])) / months
+    else:
+        m_days = (365 - sum(MONTHS_DAYS[now.month - 1 : d.month - 1])) / (12 + months)
```
Comment (carltongibson) [inline django/utils/timesince.py:79]: Can you extract this into a separate helper function, so that its purpose is clearer? Maybe a docstring? And a few unit tests.

File: django/utils/timesince.py [hunk 3]
```diff
@@ -69,22 +71,38 @@ def timesince(d, now=None, reversed=False, time_strings=None, depth=2):
...
         return avoid_wrapping(time_strings["minute"] % {"num": 0})
+
     result = []
     current_depth = 0
-    while i < len(TIMESINCE_CHUNKS) and current_depth < depth:
-        seconds, name = TIMESINCE_CHUNKS[i]
+
+    if months <= 0:
+        months = 12 + months
+
```
Comment (carltongibson) [inline django/utils/timesince.py:100]: This could be moved with the above code I think. (The `months` variable is not used in-between ln79 and here.) 🤔

If we can adjust the time chucks and the months, all in one place, with a good comment (or docstring if it can be wrapped in a function) then I think it would be clearer.

File: django/utils/timesince.py [hunk 4]
```diff
@@ -69,22 +71,38 @@ def timesince(d, now=None, reversed=False, time_strings=None, depth=2):
...
+    while i < len(timesince_chunks_adjusted) and current_depth < depth:
+        seconds, name = timesince_chunks_adjusted[i]
         count = since // seconds
+        if name == "month":
+            count = min(count, months)
```
Comment (carltongibson) [inline django/utils/timesince.py:105]: I need to have a think about this. If the timesince chucks are correct, why do we still need this? 🤔 (Is there an example?)

File: django/utils/timesince.py [hunk 5]
```diff
@@ -69,22 +71,38 @@ def timesince(d, now=None, reversed=False, time_strings=None, depth=2):
             leapdays += 1
     delta -= datetime.timedelta(leapdays)
 
+    # fix months duration to be relative to the months between d and now
+    months = now.month - d.month
+    if months > 0:
+        m_days = (sum(MONTHS_DAYS[d.month - 1 : now.month - 1])) / months
+    else:
+        m_days = (365 - sum(MONTHS_DAYS[now.month - 1 : d.month - 1])) / (12 + months)
+    timesince_chunks_adjusted = list(TIMESINCE_CHUNKS)
```
Comment (carltongibson) [inline django/utils/timesince.py:80]: This could perhaps be simply `time_chunks` 🤔

**Review by carltongibson — COMMENTED** [review]
Hi @GianpaoloBranca. Thanks for this. 

I have to say (somewhat remarkably) I **don't** dislike it 🙂 — I had thought it likely not worth the complexity, but I think the adjustment for the variable month length is probably simple enough. 

I left a couple of comments. Mainly, can we encapsulate the conversion here, so that we can test it independently, and someone opening the file for the first time can easily see that it's correct, without needing to think it through themselves each time?

I'm wondering if we can remove the need for the `months` variable beyond the block where we adjust the time since chunks? 🤔

What do you think? (I'm happy to input) 

Thanks again! ⛵️

File: django/utils/timesince.py (hunk 4 above)
Comment (GianpaoloBranca) [reply django/utils/timesince.py:105]: This was quite tricky and I discover it only when I wrote all the tests:
Consider this example:

- `d = 2022/1/1`
- `now = 2022/5/31`

let's see what the algorithm does:

1. months is 5-1 = 4
2. the average month length in this time frame is (31+28+31+30) / 4 = 30 days

if we look at the dates, between d and now there are exactly 150 days...
and if the month duration is 30 the result of `since // seconds` will be 5 instead of 4.

Without that adjustment, the algorithm might fail in cases where the average month duration between `d` and `now`
is smaller than the month duration of the month for `now`.

There are still other cases where, even with the adjustment, the result is questionable, consider this example:

- `d = 2022/1/31`
- `now = 2022/3/1`

returns `4 weeks, 1 day`...but an entire month has already passed, so `1 month` would be the correct result.
I would argue that `4 weeks, 1 day` is also correct, it is exactly 29 days, the days elapsed in the timeframe.

In conclusion, the current implementation is not yet 100% bulletproof, but it is an improvement over the previous 30 days constant.

File: django/utils/timesince.py (hunk 3 above)
Comment (GianpaoloBranca) [reply django/utils/timesince.py:100]: The reason I put this here is because the variable is not needed for the code block above, so to put it there would be confusing in my opinion. I put it close to the loop where it is used, near the other variables that are set up just before the loop. I can add a comment to make it more clear why it is there and why we still need that variable.

File: django/utils/timesince.py [hunk 6]
```diff
@@ -69,22 +71,37 @@ def timesince(d, now=None, reversed=False, time_strings=None, depth=2):
...
     result = []
     current_depth = 0
-    while i < len(TIMESINCE_CHUNKS) and current_depth < depth:
-        seconds, name = TIMESINCE_CHUNKS[i]
+
+    while i < len(time_chunks) and current_depth < depth:
+        seconds, name = time_chunks[i]
         count = since // seconds
+        if name == "month":
+            # count might overshoot if now.month is longer than the month average length
```
Comment (carltongibson) [inline django/utils/timesince.py:103]: @GianpaoloBranca Can I just clarify, did you mean `if now.month is shorter than the month average`?

File: django/utils/timesince.py (hunk 4 above)
Comment (carltongibson) [reply django/utils/timesince.py]: OK, thanks for this. 

I'm going to say we can leave the February case.

**Comment by GianpaoloBranca** [comment]
Hi @carltongibson, thanks for your comments. 

I have been working on a "bulletproof" algorithm for the month and I got it working but, let me say, it looks convoluted and ugly. Part of the reason is that I tried to fit it into the current implementation, so instead of getting directly the number of months, I have to calculate a chunk of time that when used in the main loop would produce the correct result...

I have also found other issues when working on timesince, related to leap days, because the current implementation deals with leap days by ignoring them. In the testcase [here](https://github.com/django/django/blob/main/tests/utils_tests/test_timesince.py#L41), it expect the result `one year`, but to be precise, as 2008 is a leap year and `self.oneyear` is equal to 365 days, the result should have been 11 months, 4 weeks...

But if I start taking all of this into account, it looks like there are much more things to fix than what is in the original scope of the ticket...what should I do? I can work on a reimplementation of timesince, maybe on a new ticket, or we can merge the current one as is and say that the current fix si good enough.

What do you think?

**Comment by carltongibson** [comment]
I'm just experimenting now. Give me a little while. 

Bar small edits it's likely OK as you have it. 

Thanks for the effort!

**Comment by carltongibson** [comment]
One thing that might be worth just noting it the test cases for which the current proposal is still sub optimal

**Comment by carltongibson** [comment]
Hey @GianpaoloBranca — I think this is good. 

* I wondered about extracting the adjustment logic to a helper function (not least so it can have a docstring...) 
* Also, for the `min(count, months)` — could we just use `months` there? 

Roughly this diff: 

```diff
diff --git a/django/utils/timesince.py b/django/utils/timesince.py
index 2c419d28a566e6fc61ab7a9e7b002bc3bd3ed032..47ce5324a6e1761a3fe3021a69a11c5b167b6f36 100644
--- a/django/utils/timesince.py
+++ b/django/utils/timesince.py
@@ -26,6 +26,25 @@ TIMESINCE_CHUNKS = (
 MONTHS_DAYS = (31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31)
 
 
+def get_time_chunks_and_months(now, d):
+    """
+    Return adjusted time_chucks for lengths of calendar months in interval and
+    number_of_months.
+
+    The number_of_months is needed later when now.month is longer than the count
+    given by the average.
+    """
+    number_of_months = now.month - d.month
+    if number_of_months > 0:
+        m_days = (sum(MONTHS_DAYS[d.month - 1: now.month - 1])) / number_of_months
+    else:
+        number_of_months = 12 + number_of_months
+        m_days = (365 - sum(MONTHS_DAYS[now.month - 1: d.month - 1])) / number_of_months
+    time_chunks = list(TIMESINCE_CHUNKS)
+    time_chunks[1] = (int(m_days * 24 * 60 * 60), "month")
+    return time_chunks, number_of_months
+
+
 def timesince(d, now=None, reversed=False, time_strings=None, depth=2):
     """
     Take two datetime objects and return the time between d and now as a nicely
@@ -71,15 +90,7 @@ def timesince(d, now=None, reversed=False, time_strings=None, depth=2):
             leapdays += 1
     delta -= datetime.timedelta(leapdays)
 
-    # adjust months duration to be relative to the months between d and now
-    months = now.month - d.month
-    if months > 0:
-        m_days = (sum(MONTHS_DAYS[d.month - 1 : now.month - 1])) / months
-    else:
-        months = 12 + months  # note: we also need this later to fix edge cases.
-        m_days = (365 - sum(MONTHS_DAYS[now.month - 1 : d.month - 1])) / months
-    time_chunks = list(TIMESINCE_CHUNKS)
-    time_chunks[1] = (int(m_days * 24 * 60 * 60), "month")
+    time_chunks, number_of_months = get_time_chunks_and_months(now, d)
 
     # ignore microseconds
     since = delta.days * 24 * 60 * 60 + delta.seconds
@@ -100,8 +111,8 @@ def timesince(d, now=None, reversed=False, time_strings=None, depth=2):
         seconds, name = time_chunks[i]
         count = since // seconds
         if name == "month":
-            # count might overshoot if now.month is longer than the month average length
-            count = min(count, months)
+            # ???: can we just use number_of_months here?
+            count = min(count, number_of_months)  
         if count == 0:
             break
         result.append(avoid_wrapping(time_strings[name] % {"num": count}))
```


What do you think?

**Comment by GianpaoloBranca** [comment]
Hi Carlton, I like the extraction, it looks clean, I will update the code :)

About the second point, no, we cannot use just `number_of_months`, it would fails in a lot of testcases.

Whenever `now` is in a months or more ahead of `d`, but `d.day > now.day`, it means that an entire months has not yet passed, but `number_of_months` would think so, and it result is that it is greater than what it should be.  `since // seconds` produces the corret result in this case.

For example:
- `d = 2022/05/09`
- `now = 2022/07/02`

would produce `2 months, -1 weeks` by only using `number_of_months`

**Comment by GianpaoloBranca** [comment]
Hi @carltongibson, I have been working on the optimal implementation and I think I got it! 🎉

I had to rewrite partially the whole function, and used a different strategy, so all the stuff about the average month is scrapped... but the solution looks much cleaner than the current fix

I am sorry that you have to review my code again, but I thought was better to show you the better implementation rather thain going ahead with the current one.

What do you think?

File: django/utils/timesince.py [hunk 7]
```diff
@@ -14,14 +13,16 @@
...
 
-TIMESINCE_CHUNKS = (
-    (60 * 60 * 24 * 365, "year"),
-    (60 * 60 * 24 * 30, "month"),
-    (60 * 60 * 24 * 7, "week"),
-    (60 * 60 * 24, "day"),
-    (60 * 60, "hour"),
-    (60, "minute"),
-)
+TIME_STRING_KEYS = list(TIME_STRINGS.keys())
```
Comment (carltongibson) [inline django/utils/timesince.py:16]: Does this still work correctly if I pass a custom `time_strings` to `timesince`? Could you show that with a test please?

File: django/utils/timesince.py [hunk 8]
```diff
@@ -60,37 +61,60 @@ def timesince(d, now=None, reversed=False, time_strings=None, depth=2):
...
+        if pivot_month > 12:
+            pivot_month -= 12
+            pivot_year += 1
+        pivot = datetime.datetime(
+            pivot_year,
+            pivot_month,
+            min(MONTHS_DAYS[pivot_month - 1], d.day),
+            d.hour,
+            d.minute,
+        )
```
Comment (carltongibson) [inline django/utils/timesince.py:88]: I think this could do with explaining in the docstring, which needs updating. 

(To what extent is this still based on @nedbat's original? 🤔)

**Review by carltongibson — COMMENTED** [review]
Hi @GianpaoloBranca — I've left a couple of comments: a couple of extra tests and the docstring adjustments to be going on with. 

(In the meantime, let me keep looking at it — I need to get clear on the `pivot` — but 👍)

File: django/utils/timesince.py (hunk 7 above)
Comment (GianpaoloBranca) [reply django/utils/timesince.py]: Thanks for this, I missed that point. At first I though "oh damn! I forgot about the custom `time_strings`!" And made the keys local as you suggested.

But that was also incorrect, as it would cause weird behaviors. For example, if the custom `time_strings` has the keys in a different orders.  If the given dictionary is ordered in reverse, and there is one year between the two dates, the output would actually be "one minute". 

I think that, even if it's possible to receive custom `time_strings`, it doesn't make sense for it to not have at least that set of keys. And the reason why tests were passing is that, as long as the custom `time_strings` contains the required keys, the algorithm works fine.

So yes, I can add some more tests, but what is the expected behavior in case the function receives nonsense or incomplete dictionaries? 

Currently, if the algorithm detects that there is one year between two dates, but the "year" key is not in the `time_strings` dictionary, if fails with `KeyError: 'year'`.

File: django/utils/timesince.py (hunk 8 above)
Comment (GianpaoloBranca) [reply django/utils/timesince.py]: I updated the docstring with some more explanation. I would say the original base still exists for constant time chunks, but for years and months I took some inspiration from [dateutil](https://github.com/dateutil/dateutil). The pivot is an idea of mine to make the two concept work together.

File: django/utils/timesince.py [hunk 9]
```diff
@@ -69,21 +69,23 @@ def timesince(d, now=None, reversed=False, time_strings=None, depth=2):
...
         total_months -= 1
     years, months = divmod(total_months, 12)
-    # calculate the remaining time
+    # Calculate the remaining time.
+    # Create a "pivot" datetime shifted from d by years and months, then use
+    # that to determine the other parts.
     if years or months:
-        pivot_month = d.month + months
         pivot_year = d.year + years
+        pivot_month = d.month + months
```
Comment (carltongibson) [inline django/utils/timesince.py:88]: I just swapped the variable ordering here, as it seemed clearer.

**Review by carltongibson — APPROVED** [review]
Hi @GianpaoloBranca — happy new year! 🎉 

Thanks for the work here again — it's taken a while to get there but I agree this version is good. 

I've pushed some edits in 7c33bc8744823bee299ddcbce15d1bc16659d3dd — smallest tweaks, plus adding a couple of comments to clarify the algorithm. (I imagine returning to this in five years and being glad there was an extra signpost, so future me doesn't have to think it out again 😅) 

I'll mark this _Ready for Checkin_ now. Normally we squash and rebase to a single commit on top of the current `main` — I don't know if you're about and would like to do that? Otherwise I will do so later on. (No problem either way, but I thought you might want to. 🙂) 

Welcome aboard! ⛵

**Comment by GianpaoloBranca** [comment]
Thanks a lot @carltongibson!! And happy new year to you too :)
